ssl/quic/quic_wire.c

```diff
@@ -366,18 +367,45 @@ int ossl_quic_wire_encode_frame_path_response(WPACKET *pkt,
 int ossl_quic_wire_encode_frame_conn_close(WPACKET *pkt,
                                            const OSSL_QUIC_FRAME_CONN_CLOSE *f)
 {
-    if (!encode_frame_hdr(pkt, f->is_app ? OSSL_QUIC_FRAME_TYPE_CONN_CLOSE_APP
-                                         : OSSL_QUIC_FRAME_TYPE_CONN_CLOSE_TRANSPORT)
-            || !WPACKET_quic_write_vlint(pkt, f->error_code))
+    const unsigned int is_1d = f->is_app;
+
+    /*
+     * RFC 9000 10.2.3: A CONNECTION_CLOSE of type 0x1d MUST be replaced
+     * by a CONNECTION_CLOSE of type 0x1c when sending the frame in Initial
+     * or Handshake packets.
+     */
+    if (!encode_frame_hdr(pkt, is_1d ? OSSL_QUIC_FRAME_TYPE_CONN_CLOSE_APP
+                                     : OSSL_QUIC_FRAME_TYPE_CONN_CLOSE_TRANSPORT))
         return 0;
 
-    if (!f->is_app && !WPACKET_quic_write_vlint(pkt, f->frame_type))
+    /*
+     * RFC 9000 10.2.3: Endpoints ... SHOULD use the APPLICATION_ERROR
+     * code when converting to a CONNECTION_CLOSE of type 0x1c.
+     */
+     if (!WPACKET_quic_write_vlint(pkt, is_1d ? f->error_code
+                                              : QUIC_ERR_APPLICATION_ERROR))
```

Comment:
I'm a little concerned about this change.

For a 1c close, the codes come from the space defined in [20.1](https://datatracker.ietf.org/doc/html/rfc9000#transport-error-codes) which includes the application error code.  That section says they are applicable to 1c closes, yet 1c closes strongly suggest only sending application errors.

A better approach might be converting all non-20.1 code to application error and leaving the rest intact.  I.e. leave codes 00 - 3f and 100 -1ff unchanged.

Reply:
We're not sending arbitrary error codes with 1c close. We're sending only these from 20.1.

This change does not make sense. Also please see https://github.com/openssl/openssl/pull/21135/commits/741a3abd92e120bb57f2f3bbc89bc395c5149829

Reply:
Actually we seem to be doing exactly this:

https://github.com/openssl/openssl/blob/c3a5ee3583efe5791dd4eb2827887264479a3ba2/ssl/quic/quic_wire.c#L370-L373

We're changing the packet type without guaranteeing that the error code etc are correct, although some things are taken care of.  I don't see how or why this piece of code is _safe_.